### PR TITLE
Explicit color conversion methods

### LINKED ISSUE #10319: Trait-based color conversions to/from Vec4 are too implicit and should be removed
No problem! I felt like `Vec4::from(color).truncate()` is a bit boilerplate-y, so I opened this PR.

_Originally posted by @st0rmbtw in https://github.com/bevyengine/bevy/issues/10318#issuecomment-1784969213_
            
Instead of the existing approach (and the one proposed in the linked PR), I think we should have explicit impl methods such as `Color::rgba_from_vec4(vec: Vec4)`.

It is very easy to screw up your color spaces currently, and forcing conversion methods to declare the color space used will improve matters.

### PR DESCRIPTION
**Pull request by st0rmbtw**
# Objective

Closes #10319 

## Changelog
* Added a new `Color::rgba_from_array([f32; 4]) -> Color` method.
* Added a new `Color::rgb_from_array([f32; 3]) -> Color` method.
* Added a new `Color::rgba_linear_from_array([f32; 4]) -> Color` method.
* Added a new `Color::rgb_linear_from_array([f32; 3]) -> Color` method.
* Added a new `Color::hsla_from_array([f32; 4]) -> Color` method.
* Added a new `Color::hsl_from_array([f32; 3]) -> Color` method.
* Added a new `Color::lcha_from_array([f32; 4]) -> Color` method.
* Added a new `Color::lch_from_array([f32; 3]) -> Color` method.
* Added a new `Color::rgba_to_vec4(&self) -> Vec4` method.
* Added a new `Color::rgba_to_array(&self) -> [f32; 4]` method.
* Added a new `Color::rgb_to_vec3(&self) -> Vec3` method.
* Added a new `Color::rgb_to_array(&self) -> [f32; 3]` method.
* Added a new `Color::rgba_linear_to_vec4(&self) -> Vec4` method.
* Added a new `Color::rgba_linear_to_array(&self) -> [f32; 4]` method.
* Added a new `Color::rgb_linear_to_vec3(&self) -> Vec3` method.
* Added a new `Color::rgb_linear_to_array(&self) -> [f32; 3]` method.
* Added a new `Color::hsla_to_vec4(&self) -> Vec4` method.
* Added a new `Color::hsla_to_array(&self) -> [f32; 4]` method.
* Added a new `Color::hsl_to_vec3(&self) -> Vec3` method.
* Added a new `Color::hsl_to_array(&self) -> [f32; 3]` method.
* Added a new `Color::lcha_to_vec4(&self) -> Vec4` method.
* Added a new `Color::lcha_to_array(&self) -> [f32; 4]` method.
* Added a new `Color::lch_to_vec3(&self) -> Vec3` method.
* Added a new `Color::lch_to_array(&self) -> [f32; 3]` method.

## Migration Guide
`Color::from(Vec4)` is now `Color::rgba_from_array(impl Into<[f32; 4]>)`
`Vec4::from(Color)` is now `Color::rgba_to_vec4(&self)`

Before:
```rust
let color_vec4 = Vec4::new(0.5, 0.5, 0.5);
let color_from_vec4 = Color::from(color_vec4);

let color_array = [0.5, 0.5, 0.5];
let color_from_array = Color::from(color_array);
```
After:
```rust
let color_vec4 = Vec4::new(0.5, 0.5, 0.5);
let color_from_vec4 = Color::rgba_from_array(color_vec4);

let color_array = [0.5, 0.5, 0.5];
let color_from_array = Color::rgba_from_array(color_array);
```

### CODE REVIEW

File: crates/bevy_render/src/color/mod.rs [hunk 1]
```diff
@@ -1214,48 +1368,47 @@ impl Add<Color> for Color {
 
 impl AddAssign<Vec4> for Color {
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/color/mod.rs:1369]: We should probably remove this impl as well, for the same reasons of implicitness.

**Review by alice-i-cecile — COMMENTED** [review]
Thanks for tackling this! I really like the diff on the existing code: it's much more explicit about what exactly is going on.

File: crates/bevy_render/src/color/mod.rs [hunk 2]
```diff
@@ -1083,69 +1083,224 @@ impl Color {
...
+        Color::rgba(vec4.x, vec4.y, vec4.z, vec4.w)
     }
-}
 
-impl AddAssign<Color> for Color {
-    fn add_assign(&mut self, rhs: Color) {
-        match self {
+    /// New `Color` from `[f32; 4]` in sRGB colorspace.
+    #[inline]
+    pub fn rgba_from_array([r, g, b, a]: [f32; 4]) -> Self {
```
Comment (MrGVSV) [inline crates/bevy_render/src/color/mod.rs:1095]: Would it make sense for these methods to take two parameters, with the second being the target color space? Or are we trying to keep the function signature `fn(Vec3) -> Color`?

**Comment by killercup** [comment]
I like that the color space is explicit but the input type name could be elided using custom conversion traits, which would look like `Color::rgba_from(Vec4)`, `Color::rgba_from(Vec3), and `Color::rgba_from([f32; 4])`.

**Comment by alice-i-cecile** [comment]
I like the way that looks, but it will make the docs harder to read and require importing a trait 🤔 Mildly in favor of it still I think?

**Comment by st0rmbtw** [comment]
So what we are going to do? Conversion traits or leave this as it is?

**Comment by alice-i-cecile** [comment]
@mockersf you're good at having opinions; want to make a call?

**Comment by devil-ira** [comment]
We could make the functions accept both arrays and glam's vector types like this:
```rust
fn rgba_from(values: impl Into<[f32; 4]>)
```

**Comment by killercup** [comment]
I'm sorry for starting a bikeshed 😅 

> We could make the functions accept both arrays and glam's vector types like this:
> 
> ```rust
> fn rgba_from(values: impl Into<[f32; 4]>)
> ```

This only works for Vec4

**Comment by devil-ira** [comment]
> This only works for Vec4

The `Vec3` case would be a separate method, `from_rgb` like the existing constructors.

File: crates/bevy_render/src/color/mod.rs [hunk 3]
```diff
@@ -1083,69 +1083,184 @@ impl Color {
             }
         }
     }
-}
 
-impl Default for Color {
-    fn default() -> Self {
-        Color::WHITE
+    /// New `Color` from `[f32; 4]` in sRGB colorspace.
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/color/mod.rs:1087]: These comments are now outdated.

File: crates/bevy_render/src/color/mod.rs [hunk 4]
```diff
@@ -1083,69 +1083,184 @@ impl Color {
...
+    #[inline]
+    pub fn rgba_from(arr: impl Into<[f32; 4]>) -> Self {
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/color/mod.rs:1089]: I think that the `rgb_from` is a bit awkward. IMO stick with `rgba_from_array` etc for all these method names :)

File: crates/bevy_render/src/color/mod.rs (hunk 3 above)
Comment (st0rmbtw) [reply crates/bevy_render/src/color/mod.rs:1087]: Not sure about docs. Should it be now: "New `Color` from `impl Into<[f32; 4]>` in sRGB colorspace"?

Comment (alice-i-cecile) [reply crates/bevy_render/src/color/mod.rs:1087]: "New Color from [f32; 4] (or a type that be converted into them) in sRGB colorspace"

**Comment by alice-i-cecile** [comment]
Once the docs and PR description are cleaned up this LGTM :)

**Review by alice-i-cecile — APPROVED** [review]

**Review by killercup — COMMENTED** [review]
I'd still skip the `_array` suffix and keep the `from_` prefix (`from_rgba_linear` instead `rgba_linear_from_array`) but maybe I've cause enough bike shedding 😅

**Comment by MrGVSV** [comment]
I'm still wondering if it would make sense to introduce a `ColorSpace` enum and just specify that as a second parameter so we don't have to introduce a bunch of new methods for each color space.

I suppose it would then make more sense to call it `new` or something in that case since `from` methods tend to only take a single parameter.

So something like:

```rust
let color = Color::new(my_vec4, ColorSpace::Rgba);
```

**Comment by st0rmbtw** [comment]
@MrGVSV I like the idea with the enum, but we can't have the only `new` method, because we want to create a `Color` from different types like: `Vec4`, `[f32; 4]`, `Vec3`, `[f32; 3]` and Rust doesn't allow to overload functions.

**Comment by MrGVSV** [comment]
> @MrGVSV I like the idea with the enum, but we can't have the only `new` method, because we want to create a `Color` from different types like: `Vec4`, `[f32; 4]`, `Vec3`, `[f32; 3]` and Rust doesn't allow to overload functions.

Right. In my head I was thinking `new_vec4` and such but thinking on this now I feel like keeping the `from_vec4` should be okay despite the second argument. It reads clearer at least.

I’m not saying we have to do this, just putting the idea out there in case we want to do something like it 😄

**Review by inodentry — APPROVED** [review]
I highly approve of this PR. Given that `Color` supports different color spaces, we should have API coverage of each. APIs should be explicit, not have implicit assumptions on RGB. I've seen plenty of bugs in Bevy projects due to various things just assuming RGB. The "Rusty" thing to do is to have explicit APIs so you can always tell what you are working with.

Looked through the code, did not notice anything wrong.

**Comment by alice-i-cecile** [comment]
@st0rmbtw the CI errors are real: can you fix up the broken doc tests?

**Review by alice-i-cecile — APPROVED** [review]

**Pull request merged**